agent/server.go

```diff
@@ -147,26 +178,37 @@ func newRPCServer(agent *agent, caFile, certFile, keyFile string, rpcAddrs []str
 
 	// handler is a multiplexer for both gRPC and HTTPS queries.
 	// The HTTPS endpoint returns the cluster status as JSON
+	// TODO: why does server need to handle both gRPC and HTTPS queries?
 	handler := grpcHandlerFunc(server, healthzHandler)
 
 	for _, addr := range rpcAddrs {
-		go func(address string) {
-			if err := serve(address, certFile, keyFile, tlsConfig, handler); err != nil {
-				log.WithError(err).Errorf("Failed to serve on %v.", address)
+		srv := newHTTPServer(addr, tlsConfig, handler)
+		server.httpServers = append(server.httpServers, srv)
+
+		// TODO: separate Start function to start listening.
+		go func(srv *http.Server) {
+			err := srv.ListenAndServeTLS(certFile, keyFile)
+			if err == http.ErrServerClosed {
+				log.WithError(err).Debug("Server has been shutdown/closed.")
+				return
+			}
+			if err != nil {
+				log.WithError(err).Errorf("Failed to serve on %v.", srv.Addr)
```

Comment:
this block would ideally trigger agent process to fail as well.

Reply:
I was thinking it would be better to move this block into a separate `start` function.
We can then implement some sort of retry strategy and either wait and return any errors or return an error channel for the caller to wait to handle server shutdown.